File: macos/SeedSeeker/Sources/SeedSeekerKit/ResultsExport.swift
import CSeedFinder
import Foundation

/// User-facing failure while reading a results file.
public struct ResultsExportError: Error, LocalizedError, Equatable {
    public let message: String
    public init(_ message: String) { self.message = message }
    public var errorDescription: String? { message }
}

/// The cross-platform results-export document: search results plus the query
/// that found them.
///
/// The format is the Rust core's (`crates/seedfinder-core/src/results_export.rs`,
/// documented in `docs/results-export-format.md`) and so is every rule about
/// it: the envelope, the compatibility contract, the strict query validation,
/// the canonical seed codes, the import size cap and the dedupe-and-cap step
/// all live behind `seedfinder_results_encode`/`_decode`. What remains here is
/// the mapping between that canonical query document and the Swift models,
/// which the engine has already validated by the time it is read.
public enum ResultsExport {
    public static let suggestedFileName = "seed-seeker-results"

    public struct Imported: Sendable {
        public let query: SavedQuery
        public let seeds: [String]
        /// Exported entries the engine's dedupe-and-cap step removed.
        public let dropped: Int
        /// The upstream game version the file declares, if any.
        public let shpdVersion: String?
        public init(query: SavedQuery, seeds: [String], dropped: Int, shpdVersion: String?) {
            self.query = query; self.seeds = seeds
            self.dropped = dropped; self.shpdVersion = shpdVersion
        }
    }

    /// Stable document names, indexed by the matching enum raw value.
    private static let kindNames = ["weapon", "armor", "wand", "ring", "melee_weapon", "thrown_weapon", "trinket", "artifact"]
    private static let sourceNames = [
        "heap", "chest", "locked_chest", "crystal_chest", "tomb", "skeleton",
        "sacrificial_fire", "mimic", "golden_mimic", "crystal_mimic", "statue",
        "armored_statue", "shop", "ghost_reward", "wandmaker_reward",
        "blacksmith_reward", "imp_reward", "vault_treasure",
    ]
    private static let challengeNames: [(name: String, challenge: Challenge)] = [
        ("on_diet", .noFood), ("faith_is_my_armor", .noArmor),
        ("pharmacophobia", .noHealing), ("barren_land", .noHerbalism),
        ("swarm_intelligence", .swarmIntelligence), ("into_darkness", .darkness),
        ("forbidden_runes", .noScrolls), ("hostile_champions", .championEnemies),
        ("badder_bosses", .strongerBosses),
    ]

    /// Encodes the query and its seeds as results-file text, or "" when the
    /// engine refuses them (an unusable query, a non-canonical seed code).
    public static func encode(_ query: SavedQuery, seeds: [String], appVersion: String) -> String {
        let request: [String: Any] = [
            "query": encodeQuery(query),
            "seeds": seeds,
            "app_version": appVersion,
        ]
        guard let document = try? JSONSerialization.data(withJSONObject: request),
              let packet = try? enginePacket({ out, length in
                  document.withUnsafeBytes { bytes in
                      seedfinder_results_encode(
                          bytes.bindMemory(to: UInt8.self).baseAddress, bytes.count, out, length)
                  }
              }),
              let text = String(data: packet, encoding: .utf8) else { return "" }
        return text
    }

    public static func decode(_ text: String) throws -> Imported {
        let packet: Data
        do {
            packet = try enginePacket { out, length in
                Data(text.utf8).withUnsafeBytes { bytes in
                    seedfinder_results_decode(
                        bytes.bindMemory(to: UInt8.self).baseAddress, bytes.count, out, length)
                }
            }
        } catch SeedFinderEngineError.invalidArgument {
            throw ResultsExportError("This is not a Seed Seeker results file this version can import.")
        } catch {
            throw ResultsExportError("The native engine failed while reading the results file.")
        }
        guard let document = (try? JSONSerialization.jsonObject(with: packet)) as? [String: Any],
              let queryValue = document["query"] as? [String: Any],
              let seeds = document["seeds"] as? [String] else {
            throw ResultsExportError("The native engine returned an invalid results document.")
        }
        return Imported(query: try decodeQuery(queryValue), seeds: seeds,
                        dropped: intField(document, "dropped") ?? 0,
                        shpdVersion: document["shpd_version"] as? String)
    }

    // MARK: Document mapping

    // The document the engine hands back has already been validated by it, so
    // these readers only translate: a value of an unexpected shape simply is
    // not there.
    private static func intField(_ entry: [String: Any], _ key: String) -> Int? {
        guard let number = entry[key] as? NSNumber else { return nil }
        return Int(exactly: number)
    }

    private static func boolField(_ entry: [String: Any], _ key: String) -> Bool {
        (entry[key] as? NSNumber)?.boolValue ?? false
    }

    // Internal, not private: the share-link codec (`DeepLink`) and the engine
    // transport (`QueryDocument`) exchange the same canonical query document
    // with the Rust core.
    static func encodeQuery(_ query: SavedQuery) -> [String: Any] {
        // An alternative group is one `any_of` entry at its first member's
        // position, members in requirement order; a lone requirement is written plain.
        let entries: [Any] = query.requirements.slots.map { slot in
            slot.count == 1 ? encodeRequirement(slot[0]) : ["any_of": slot.map(encodeRequirement)]
        }
        var output: [String: Any] = ["requirements": entries]
        if query.maximumDepth != 24 { output["max_depth"] = query.maximumDepth }
        if query.requireBlacksmith { output["require_blacksmith"] = true }
        if query.excludeBlacksmithRewards { output["exclude_blacksmith_rewards"] = true }
        if let quest = query.wandmakerQuest { output["wandmaker_quest"] = quest.documentName }
        let challenges = challengeNames
            .filter { query.challenges & $0.challenge.rawValue != 0 }
            .map(\.name)
        if !challenges.isEmpty { output["challenges"] = challenges }
        return output
    }

    private static func encodeRequirement(_ requirement: ItemRequirement) -> [String: Any] {
        var output: [String: Any] = ["kind": kindNames[requirement.kind.rawValue]]
        if let item = requirement.item { output["item"] = item.id }
        switch requirement.tierMatch {
        case .any: break
        case .exactly: output["tier"] = ["exact": requirement.tier]
        case .atLeast: output["tier"] = ["at_least": requirement.tier]
        case .atMost: output["tier"] = ["at_most": requirement.tier]
        }
        switch requirement.upgradeMatch {
        case .any: break
        case .exactly: output["upgrade"] = requirement.upgrade
        case .atLeast: output["upgrade"] = ["at_least": requirement.upgrade]
        }
        switch requirement.effect {
        case .any: break
        case .anyEnchantment: output["effect"] = anyEnchantmentName
        case .oneOf(let names): output["effect"] = names.count == 1 ? names[0] : names
        }
        if requirement.requireUncursed { output["uncursed"] = true }
        if let source = requirement.source { output["source"] = sourceNames[source.rawValue] }
        if let group = requirement.identityGroup { output["identity_group"] = group }
        if let depth = requirement.maximumDepth { output["max_depth"] = depth }
        if let sum = requirement.levelSum { output["level_sum"] = ["group": sum.group, "at_least": sum.atLeast] }
        return output
    }

    /// The document's shorthand for the family's whole non-curse effect set.
    private static let anyEnchantmentName = "any_enchantment"

    /// Maps a canonical query document onto the Swift models. Only a name this
    /// build has no model for fails — the document itself is the engine's, so
    /// its shape and bounds are already guaranteed.
    static func decodeQuery(_ value: [String: Any]) throws -> SavedQuery {
        var requirements: [ItemRequirement] = []
        // Alternative groups get fresh sequential ids in document order.
        var nextGroup = 1
        for (index, entry) in (value["requirements"] as? [Any] ?? []).enumerated() {
            guard let entry = entry as? [String: Any] else {
                throw ResultsExportError("Requirement \(index + 1) is not a JSON object.")
            }
            do {
                if let members = entry["any_of"] as? [Any] {
                    let group = nextGroup
                    nextGroup += 1
                    for member in members {
                        guard let member = member as? [String: Any] else {
                            throw ResultsExportError("an alternative is not a JSON object")
                        }
                        requirements.append(try decodeRequirement(
                            member, key: Int64(requirements.count + 1), alternativeGroup: group))
                    }
                } else {
                    requirements.append(try decodeRequirement(entry, key: Int64(requirements.count + 1)))
                }
            } catch let failure as ResultsExportError {
                throw ResultsExportError("Requirement \(index + 1): \(failure.message)")
            } catch {
                let reason = (error as? LocalizedError)?.errorDescription ?? "\(error)"
                throw ResultsExportError("Requirement \(index + 1): \(reason)")
            }
        }
        var challenges = 0
        for name in value["challenges"] as? [String] ?? [] {
            guard let match = challengeNames.first(where: { $0.name == name }) else {
                throw ResultsExportError("Unknown challenge \"\(name)\".")
            }
            challenges |= match.challenge.rawValue
        }
        var wandmakerQuest: WandmakerQuest?
        if let name = value["wandmaker_quest"] as? String {
            guard let quest = WandmakerQuest.named(name) else {
                throw ResultsExportError("Unknown Wandmaker quest \"\(name)\".")
            }
            wandmakerQuest = quest
        }
        // A `fast_mode` key from a document written before the flag was
        // retired is read past and ignored, exactly as the engine's own
        // decoder does, so those files and links still open.
        return SavedQuery(
            requirements: requirements,
            maximumDepth: intField(value, "max_depth") ?? 24,
            requireBlacksmith: boolField(value, "require_blacksmith"),
            excludeBlacksmithRewards: boolField(value, "exclude_blacksmith_rewards"),
            wandmakerQuest: wandmakerQuest,
            challenges: challenges)
    }

    private static func decodeRequirement(_ entry: [String: Any], key: Int64,
                                          alternativeGroup: Int? = nil) throws -> ItemRequirement {
        var item: CatalogItem?
        if let id = entry["item"] as? String {
            guard let found = ItemCatalog.findById(id) else {
                throw ResultsExportError("unknown item \"\(id)\"")
            }
            item = found
        }
        let kind: ItemKind
        if let name = entry["kind"] as? String {
            guard let index = kindNames.firstIndex(of: name),
                  let value = ItemKind(rawValue: index) else {
                throw ResultsExportError("unknown category \"\(name)\"")
            }
            kind = value
        } else if let item {
            kind = item.kind
        } else {
            throw ResultsExportError("a category is required when no item is set")
        }
        var tier = 0
        var tierMatch = TierMatch.any
        if let object = entry["tier"] as? [String: Any] {
            if let exact = intField(object, "exact") { tier = exact; tierMatch = .exactly }
            else if let atLeast = intField(object, "at_least") { tier = atLeast; tierMatch = .atLeast }
            else if let atMost = intField(object, "at_most") { tier = atMost; tierMatch = .atMost }
        }
        var upgrade = 0
        var upgradeMatch = UpgradeMatch.any
        if let object = entry["upgrade"] as? [String: Any] {
            if let exact = intField(object, "exact") { upgrade = exact; upgradeMatch = .exactly }
            else if let atLeast = intField(object, "at_least") { upgrade = atLeast; upgradeMatch = .atLeast }
        } else if let exact = intField(entry, "upgrade") {
            upgrade = exact
            upgradeMatch = .exactly
        }
        // Effect names match case-insensitively and canonicalize to the
        // catalog's own spelling.
        func effectName(_ name: String) throws -> String {
            guard let match = ItemCatalog.modifiersFor(kind)
                .first(where: { $0.caseInsensitiveCompare(name) == .orderedSame }) else {
                throw ResultsExportError("unknown effect \"\(name)\"")
            }
            return match
        }
        var effect = EffectFilter.any
        if let name = entry["effect"] as? String {
            effect = name.caseInsensitiveCompare(anyEnchantmentName) == .orderedSame
                ? .anyEnchantment : .oneOf([try effectName(name)])
        } else if let names = entry["effect"] as? [String] {
            effect = .oneOf(try names.map(effectName))
        }
        // The unreleased upgrade_sum key counted upgrades, not levels: refused
        // rather than silently reinterpreted, as the engine does.
        guard entry["upgrade_sum"] == nil else {
            throw ResultsExportError("upgrade_sum is no longer supported; use level_sum")
        }
        var levelSum: LevelSum?
        if let object = entry["level_sum"] as? [String: Any],
           let group = intField(object, "group"), let atLeast = intField(object, "at_least") {
            levelSum = LevelSum(group: group, atLeast: atLeast)
        }
        var source: ScoutItemSource?
        if let name = entry["source"] as? String {
            guard let index = sourceNames.firstIndex(of: name),
                  let value = ScoutItemSource(rawValue: index) else {
                throw ResultsExportError("unknown source \"\(name)\"")
            }
            source = value
        }
        return try ItemRequirement(
            key: key,
            item: item,
            upgrade: upgrade,
            effect: effect,
            kind: kind,
            tier: tier,
            tierMatch: tierMatch,
            upgradeMatch: upgradeMatch,
            source: source,
            identityGroup: intField(entry, "identity_group"),
            maximumDepth: intField(entry, "max_depth"),
            requireUncursed: boolField(entry, "uncursed"),
            alternativeGroup: alternativeGroup,
            levelSum: levelSum)
    }
}

/// The canonical JSON query document as the bytes every query-taking engine
/// entry point accepts: the search, the resumed search, the seed filter, the
/// continuation and start decisions and the scout marks all read the same
/// document the share links and results files carry.
public enum QueryDocument {
    /// The document as a JSON object, before serialization.
    public static func object(_ request: SearchRequest) -> [String: Any] {
        ResultsExport.encodeQuery(SavedQuery(
            requirements: request.requirements, maximumDepth: request.maximumDepth,
            requireBlacksmith: request.requireBlacksmith,
            excludeBlacksmithRewards: request.excludeBlacksmithRewards,
            wandmakerQuest: request.wandmakerQuest,
            challenges: request.challenges))
    }

    /// UTF-8 JSON bytes of the document, keys sorted so equal queries encode
    /// to equal bytes.
    public static func encode(_ request: SearchRequest) throws -> Data {
        try JSONSerialization.data(withJSONObject: object(request), options: [.sortedKeys])
    }
}

File: windows/SeedSeeker/ResultsExport.cs
using System.Collections.ObjectModel;
using System.Text.Json.Nodes;

namespace SeedSeeker;

/// <summary>User-facing failure while reading a results file.</summary>
public sealed class ResultsExportException(string message) : Exception(message);

/// <summary>
/// The cross-platform results-export document: search results plus the query
/// that found them.
///
/// The codec itself is the engine's — <c>seedfinder_results_encode</c> and
/// <c>seedfinder_results_decode</c> over the Rust core
/// (crates/seedfinder-core/src/results_export.rs; the schema is documented in
/// docs/results-export-format.md). It owns the envelope, the compatibility
/// rules, the 2 MiB import cap, the dedupe-and-cap step and every query
/// validation, so this file is only the mapping between the canonical query
/// document and <see cref="QuerySettings"/>. A document that arrives from the
/// engine has already been validated; the mapping is therefore lenient and
/// fails only on content it genuinely cannot represent.
/// </summary>
public static class ResultsExport
{
    public const string SuggestedFileName = "seed-seeker-results";

    /// <param name="Dropped">Exported entries the engine's dedupe-and-cap step removed.</param>
    public sealed record Imported(QuerySettings Query, IReadOnlyList<string> Seeds, int Dropped, string? FileShpdVersion);

    /// <summary>Stable document names, indexed by the matching enum value.</summary>
    private static readonly string[] KindNames = ["weapon", "armor", "wand", "ring", "melee_weapon", "thrown_weapon", "trinket", "artifact"];
    private static readonly string[] SourceNames = [
        "heap", "chest", "locked_chest", "crystal_chest", "tomb", "skeleton",
        "sacrificial_fire", "mimic", "golden_mimic", "crystal_mimic", "statue",
        "armored_statue", "shop", "ghost_reward", "wandmaker_reward",
        "blacksmith_reward", "imp_reward", "vault_treasure",
    ];
    private static readonly (string Name, int Bit)[] ChallengeNames =
        [.. Challenges.All.Select(entry => (entry.Name, entry.Mask))];

    /// <exception cref="ResultsExportException">With a user-facing message.</exception>
    public static string Encode(QuerySettings query, IEnumerable<string> seeds, string appVersion)
    {
        var request = new JsonObject
        {
            ["query"] = EncodeQuery(query),
            ["seeds"] = new JsonArray([.. seeds.Select(seed => (JsonNode)seed)]),
            ["app_version"] = appVersion,
        };
        return NativeEngine.TryEncodeResultsFile(request.ToJsonString())
            ?? throw new ResultsExportException("These results could not be written to a results file.");
    }

    /// <exception cref="ResultsExportException">With a user-facing message.</exception>
    public static Imported Decode(string text)
    {
        var decoded = NativeEngine.TryDecodeResultsFile(text)
            ?? throw new ResultsExportException(
                "This is not a Seed Seeker results file, or its query is not one this version can run.");
        if (JsonNode.Parse(decoded) is not JsonObject document || document["query"] is not JsonObject queryValue)
            throw new ResultsExportException("This results file could not be read.");
        var seeds = new List<string>();
        foreach (var entry in document["seeds"] as JsonArray ?? [])
            if (entry is JsonValue seedValue && seedValue.TryGetValue(out string? seed)) seeds.Add(seed);
        return new Imported(DecodeQuery(queryValue), seeds, IntField(document, "dropped") ?? 0,
            TolerantString(document, "shpd_version"));
    }

    /// <summary>Reads informational envelope strings; wrong types are ignored, not errors.</summary>
    private static string? TolerantString(JsonObject document, string key) =>
        document[key] is JsonValue value && value.TryGetValue(out string? text) ? text : null;

    /// <summary>The bare canonical JSON query document, the share-link codec's input.</summary>
    public static string EncodeQueryDocument(QuerySettings query) => EncodeQuery(query).ToJsonString();

    /// <summary>
    /// Decodes a bare canonical JSON query document, the share-link codec's
    /// output — which the engine has already validated.
    /// </summary>
    /// <exception cref="ResultsExportException">With a user-facing message.</exception>
    public static QuerySettings DecodeQueryDocument(string text)
    {
        if (JsonNode.Parse(text) is not JsonObject document)
            throw new ResultsExportException("The shared query could not be read.");
        return DecodeQuery(document);
    }

    /// <summary>The document's name for the full non-curse effect set of a family.</summary>
    private const string AnyEnchantment = "any_enchantment";

    private static JsonObject EncodeQuery(QuerySettings query)
    {
        // An "any of these" slot is one any_of entry at its first member's
        // position, members in requirement order; a lone member is written
        // plain, exactly as the core's writer does it.
        var entries = QueryRelationships.Slots(query.Requirements).Select(slot => slot.Count == 1
            ? (JsonNode)EncodeRequirement(slot[0])
            : new JsonObject { ["any_of"] = new JsonArray([.. slot.Select(member => (JsonNode)EncodeRequirement(member))]) });
        var output = new JsonObject { ["requirements"] = new JsonArray([.. entries]) };
        if (query.MaximumDepth != 24) output["max_depth"] = query.MaximumDepth;
        if (query.RequireBlacksmith) output["require_blacksmith"] = true;
        if (query.ExcludeBlacksmithRewards) output["exclude_blacksmith_rewards"] = true;
        if (WandmakerQuests.DocumentName(query.WandmakerQuest) is string quest) output["wandmaker_quest"] = quest;
        var challenges = ChallengeNames.Where(c => (query.Challenges & c.Bit) != 0).Select(c => c.Name).ToArray();
        if (challenges.Length != 0)
            output["challenges"] = new JsonArray([.. challenges.Select(name => (JsonNode)name)]);
        return output;
    }

    private static JsonObject EncodeRequirement(ItemRequirement requirement)
    {
        var output = new JsonObject { ["kind"] = KindNames[(int)requirement.Kind] };
        if (requirement.Item is not null) output["item"] = requirement.Item.Id;
        output["tier"] = requirement.TierMatch switch
        {
            TierMatch.Exactly => new JsonObject { ["exact"] = requirement.Tier },
            TierMatch.AtLeast => new JsonObject { ["at_least"] = requirement.Tier },
            TierMatch.AtMost => new JsonObject { ["at_most"] = requirement.Tier },
            _ => null,
        };
        if (output["tier"] is null) output.Remove("tier");
        output["upgrade"] = requirement.UpgradeMatch switch
        {
            UpgradeMatch.Exactly => JsonValue.Create(requirement.Upgrade),
            UpgradeMatch.AtLeast => new JsonObject { ["at_least"] = requirement.Upgrade },
            _ => null,
        };
        if (output["upgrade"] is null) output.Remove("upgrade");
        if (EncodeEffect(requirement) is JsonNode effect) output["effect"] = effect;
        if (requirement.RequireUncursed) output["uncursed"] = true;
        if (requirement.Source is ScoutItemSource source) output["source"] = SourceNames[(int)source];
        if (requirement.IdentityGroup is int group) output["identity_group"] = group;
        if (requirement.MaximumDepth is int depth) output["max_depth"] = depth;
        if (requirement.LevelSum is { } sum) output["level_sum"] = new JsonObject { ["group"] = sum.Group, ["at_least"] = sum.AtLeast };
        return output;
    }

    /// <summary>
    /// The writer's effect rules: the whole non-curse family set is the
    /// shorthand, one effect a bare name, anything else the names in catalog
    /// order. Null when the requirement accepts any effect.
    /// </summary>
    private static JsonNode? EncodeEffect(ItemRequirement requirement)
    {
        var filter = requirement.Effect;
        if (filter.AnyEnchantment || filter.IsEveryEnchantmentOf(requirement.Kind)) return AnyEnchantment;
        var names = ItemCatalog.Modifiers(requirement.Kind).Where(filter.Effects.Contains).ToList();
        return names.Count switch
        {
            0 => null,
            1 => names[0],
            _ => new JsonArray([.. names.Select(name => (JsonNode)name)]),
        };
    }

    private static QuerySettings DecodeQuery(JsonObject value)
    {
        var requirements = new ObservableCollection<ItemRequirement>(); var alternativeGroups = 0;
        foreach (var entry in value["requirements"] as JsonArray ?? [])
        {
            if (entry is not JsonObject requirement) continue;
            // Readers give any_of groups fresh sequential ids in document order.
            if (requirement["any_of"] is JsonArray members)
            {
                alternativeGroups++;
                foreach (var member in members)
                    if (member is JsonObject alternative) requirements.Add(DecodeRequirement(alternative, alternativeGroups));
            }
            else requirements.Add(DecodeRequirement(requirement, null));
        }
        var challenges = 0;
        foreach (var nameValue in value["challenges"] as JsonArray ?? [])
        {
            string? name = null;
            if (nameValue is JsonValue nameJson) nameJson.TryGetValue(out name);
            var match = ChallengeNames.FirstOrDefault(c => c.Name == name);
            if (match.Name is not null) challenges |= match.Bit;
        }
        return new QuerySettings
        {
            Requirements = requirements,
            MaximumDepth = IntField(value, "max_depth") ?? 24,
            RequireBlacksmith = BoolField(value, "require_blacksmith"),
            ExcludeBlacksmithRewards = BoolField(value, "exclude_blacksmith_rewards"),
            WandmakerQuest = TolerantString(value, "wandmaker_quest") is string questName
                ? WandmakerQuests.Named(questName) ?? WandmakerQuest.Any
                : WandmakerQuest.Any,
            // A "fast_mode" flag written before that mode was retired is simply
            // not read here, matching the engine's own decoder: documents that
            // carry it still load, as an ordinary full-depth search.
            Challenges = challenges,
        };
    }

    /// <exception cref="ResultsExportException">
    /// When the requirement names catalog content this build does not know —
    /// the one failure the engine cannot rule out, since the item and effect
    /// tables are the app's own copies of the shared catalog asset.
    /// </exception>
    private static ItemRequirement DecodeRequirement(JsonObject entry, int? alternativeGroup)
    {
        CatalogItem? item = null;
        if (TolerantString(entry, "item") is string id)
            item = ItemCatalog.Find(id) ?? throw new ResultsExportException($"This query names an unknown item \"{id}\".");
        ItemKind kind;
        if (TolerantString(entry, "kind") is string kindName)
        {
            var index = Array.IndexOf(KindNames, kindName);
            if (index < 0) throw new ResultsExportException($"This query names an unknown category \"{kindName}\".");
            kind = (ItemKind)index;
        }
        else
        {
            kind = item?.Kind ?? ItemKind.Weapon;
        }
        var effect = DecodeEffect(entry["effect"], kind);
        ScoutItemSource? source = null;
        if (TolerantString(entry, "source") is string sourceName)
        {
            var index = Array.IndexOf(SourceNames, sourceName);
            if (index < 0) throw new ResultsExportException($"This query names an unknown source \"{sourceName}\".");
            source = (ScoutItemSource)index;
        }
        var (tier, tierMatch) = DecodeTier(entry["tier"]);
        var (upgrade, upgradeMatch) = DecodeUpgrade(entry["upgrade"]);
        // The unreleased upgrade_sum key counted upgrades, not levels: refused
        // rather than silently reinterpreted, as the engine does.
        if (entry["upgrade_sum"] is not null)
            throw new ResultsExportException("This query uses upgrade_sum, which is no longer supported; use level_sum.");
        LevelSum? levelSum = null;
        if (entry["level_sum"] is JsonObject sum && IntField(sum, "group") is int sumGroup && IntField(sum, "at_least") is int atLeast)
            levelSum = new(sumGroup, atLeast);
        return new ItemRequirement
        {
            Item = item,
            Upgrade = upgrade,
            Effect = effect,
            Kind = kind,
            Tier = tier,
            TierMatch = tierMatch,
            UpgradeMatch = upgradeMatch,
            Source = source,
            IdentityGroup = IntField(entry, "identity_group"),
            MaximumDepth = IntField(entry, "max_depth"),
            RequireUncursed = BoolField(entry, "uncursed"),
            AlternativeGroup = alternativeGroup,
            LevelSum = levelSum,
        };
    }

    /// <exception cref="ResultsExportException">When an effect name is not in this build's catalog tables.</exception>
    private static EffectFilter DecodeEffect(JsonNode? value, ItemKind kind)
    {
        string Known(string name) => ItemCatalog.Modifiers(kind)
            .FirstOrDefault(known => string.Equals(known, name, StringComparison.OrdinalIgnoreCase))
            ?? throw new ResultsExportException($"This query names an unknown effect \"{name}\".");
        switch (value)
        {
            case JsonValue single when single.TryGetValue(out string? name):
                // The shorthand is matched case-insensitively, like the engine does.
                return string.Equals(name, AnyEnchantment, StringComparison.OrdinalIgnoreCase)
                    ? EffectFilter.Enchantment()
                    : EffectFilter.OneOf([Known(name)]);
            case JsonArray names:
                var effects = new List<string>();
                foreach (var entry in names)
                    if (entry is JsonValue nameValue && nameValue.TryGetValue(out string? name)) effects.Add(Known(name));
                // An empty list is rejected by the engine, so refuse it here too
                // rather than silently widening the requirement to "any effect".
                if (effects.Count == 0) throw new ResultsExportException("This query has an empty effect list.");
                return EffectFilter.OneOf(effects.Distinct());
            default:
                return EffectFilter.Any();
        }
    }

    private static (int, TierMatch) DecodeTier(JsonNode? value) => value switch
    {
        JsonObject filter when IntField(filter, "exact") is int exact => (exact, TierMatch.Exactly),
        JsonObject filter when IntField(filter, "at_least") is int atLeast => (atLeast, TierMatch.AtLeast),
        JsonObject filter when IntField(filter, "at_most") is int atMost => (atMost, TierMatch.AtMost),
        _ => (0, TierMatch.Any),
    };

    private static (int, UpgradeMatch) DecodeUpgrade(JsonNode? value) => value switch
    {
        JsonValue number when number.TryGetValue(out int upgrade) => (upgrade, UpgradeMatch.Exactly),
        JsonObject filter when IntField(filter, "exact") is int exact => (exact, UpgradeMatch.Exactly),
        JsonObject filter when IntField(filter, "at_least") is int atLeast => (atLeast, UpgradeMatch.AtLeast),
        _ => (0, UpgradeMatch.Any),
    };

    /// <summary>Reads a whole number; absent or wrong-typed values read as absent.</summary>
    private static int? IntField(JsonObject entry, string key) =>
        entry[key] is JsonValue value && value.TryGetValue(out int number) ? number : null;

    /// <summary>Reads a flag; absent or wrong-typed values read as false.</summary>
    private static bool BoolField(JsonObject entry, string key) =>
        entry[key] is JsonValue value && value.TryGetValue(out bool flag) && flag;
}
